File: ios/MidObjcSample/MidObjcSample/ViewController.m
//
//  ViewController.m
//  MidObjcSample
//
//  Created by Muhammad Fauzi Masykur on 07/08/20.
//  Copyright © 2020 Muhammad Fauzi Masykur. All rights reserved.
//

#import "ViewController.h"
#import <MidtransKit/MidtransKit.h>
#import "ResultViewController.h"

@interface ViewController () <MidtransUIPaymentViewControllerDelegate>

@property (weak, nonatomic) IBOutlet UITextField *snaptokenTextField;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    //setup callback url, to enable customer to go back to this app after completing payment in e-wallet app (gojek/shopeepay).
    //setup gopay callback url
    [MidtransConfig shared].callbackSchemeURL = @"sampleapp-objc://";
    //setup shopeepay callback url
    [MidtransConfig shared].shopeePayCallbackURL = @"sampleapp-objc://";
}

- (NSString *)generateRandomOrderId{
    NSDate * now = [NSDate date];
    NSDateFormatter *outputFormatter = [[NSDateFormatter alloc] init];
    [outputFormatter setDateFormat:@"HHmmss"];
    NSString *randomOrderIdFromDate = [outputFormatter stringFromDate:now];
    return randomOrderIdFromDate;
}

- (void)showSnaptokenEmptyAlert{
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Snaptoken is empty" message:@"please input a snaptoken to use a snaptoken payment" preferredStyle:UIAlertControllerStyleAlert];
    [alert addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleCancel handler:nil]];
    [self presentViewController:alert animated:YES completion:nil];
}

- (void)showPaymentStatusPage:(NSString *)paymentStatus{
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Main" bundle:nil];
    ResultViewController * vc = [storyboard instantiateViewControllerWithIdentifier:@"resultVC"];
    vc.paymentStatus = paymentStatus;
    [self.navigationController pushViewController:vc animated:YES];
}

- (IBAction)payWithSnaptokenFlow:(id)sender {
    if (self.snaptokenTextField.text.length != 0) {
        [[MidtransMerchantClient shared]requestTransacationWithCurrentToken:self.snaptokenTextField.text completion:^(MidtransTransactionTokenResponse * _Nullable response, NSError * _Nullable error) {
            if (response) {
                //present payment page
                MidtransUIPaymentViewController *paymentVC = [[MidtransUIPaymentViewController alloc] initWithToken:response];
                paymentVC.paymentDelegate = self;
                [self presentViewController:paymentVC animated:YES completion:nil];
            } else {
                NSLog(@"error %@", error);
            }
        }];
    } else {
        [self showSnaptokenEmptyAlert];
    }
}
- (IBAction)payWithSDKFlow:(id)sender {
    
    //Create Transaction Details
    MidtransTransactionDetails *transactionDetails = [[MidtransTransactionDetails alloc]initWithOrderID:[self generateRandomOrderId] andGrossAmount:@300];
    MidtransItemDetail *itemJacket = [[MidtransItemDetail alloc]initWithItemID:@"item1" name:@"jacket" price:@150 quantity:@1];
    MidtransItemDetail *itemPants = [[MidtransItemDetail alloc]initWithItemID:@"item2" name:@"pants" price:@150 quantity:@1];
    MidtransAddress *address = [MidtransAddress addressWithFirstName:@"john" lastName:@"doe" phone:@"082221009999" address:@"jl.buntu 2" city:@"Jakarta" postalCode:@"112233" countryCode:@"IDN"];
    MidtransCustomerDetails *customerDetails = [[MidtransCustomerDetails alloc]initWithFirstName:@"John" lastName:@"Doe" email:@"cihuy@cihuyhuy.com" phone:@"082221009999" shippingAddress:address billingAddress:address];
    
    //request transaction token
    [[MidtransMerchantClient shared] requestTransactionTokenWithTransactionDetails:transactionDetails itemDetails:@[itemJacket,itemPants] customerDetails:customerDetails completion:^(MidtransTransactionTokenResponse * _Nullable token, NSError * _Nullable error) {
        
        if (token) {
        //present payment page
        MidtransUIPaymentViewController *paymentVC = [[MidtransUIPaymentViewController alloc] initWithToken:token];
        paymentVC.paymentDelegate = self;
        [self presentViewController:paymentVC animated:YES completion:nil];
        } else {
            NSLog(@"error %@", error);
        }
    }];
}

//PRAGMA MARK: - MidtransUIPaymentViewControllerDelegate
//Transaction callback/delegates to handle all the transaction status
- (void)paymentViewController:(MidtransUIPaymentViewController *)viewController paymentFailed:(NSError *)error {
    NSLog(@"payment failed");
    [self showPaymentStatusPage:@"error"];
}

- (void)paymentViewController:(MidtransUIPaymentViewController *)viewController paymentPending:(MidtransTransactionResult *)result {
    NSLog(@"payment pending");
    [self showPaymentStatusPage:result.transactionStatus];
}

- (void)paymentViewController:(MidtransUIPaymentViewController *)viewController paymentSuccess:(MidtransTransactionResult *)result {
    NSLog(@"payment success");
    [self showPaymentStatusPage:result.transactionStatus];
}

- (void)paymentViewController_paymentCanceled:(MidtransUIPaymentViewController *)viewController {
    NSLog(@"payment cancelled");
    [self showPaymentStatusPage:@"cancelled"];
}
-(void)paymentViewController:(MidtransUIPaymentViewController *)viewController paymentDeny:(MidtransTransactionResult *)result {
    NSLog(@"payment denied");
    [self showPaymentStatusPage:result.transactionStatus];
}

@end

File: ios/MidtransSpmSample/MidtransSpmSample/ViewController.swift
//
//  ViewController.swift
//  xc-sample-swift
//
//  Created by Muhammad Fauzi Masykur on 26/08/22.
//

import UIKit
import MidtransKit

class ViewController: UIViewController, MidtransUIPaymentViewControllerDelegate {
    @IBOutlet weak var snaptokenTextField: UITextField!
    override func viewDidLoad() {
        super.viewDidLoad()

        //setup callback url, to enable customer to go back to this app after completing payment in e-wallet app (gojek/shopeepay).
        //setup gopay callback url
        MidtransConfig.shared()?.callbackSchemeURL = "sampleapp://"
        //setup shopeepay callback url
        MidtransConfig.shared()?.shopeePayCallbackURL = "sampleapp://"
        MidtransCreditCardConfig.shared()?.authenticationType = .type3DS
        
        
    }
    
    func generateRandomOrderID()-> String {
        let date = Date()
        let df = DateFormatter()
        df.dateFormat = "HHmmss"
        let randomOrderID = df.string(from: date)
        return randomOrderID
    }
    
    func showSnaptokenEmptyAlert() {
        let alert = UIAlertController(title: "Snaptoken is empty", message: "please input a snaptoken to use a snaptokenflow payment", preferredStyle: .alert)
        alert.addAction(UIAlertAction(title: "OK", style: .cancel, handler: nil))
        self.present(alert, animated: true)
    }
    
    func showPaymentStatusPage(paymentStatus:String){
        let storyboard = UIStoryboard(name: "Main", bundle: nil)
        let vc = storyboard.instantiateViewController(withIdentifier: "resultVC") as! ResultViewController
        vc.paymentStatus = paymentStatus
        self.navigationController?.pushViewController(vc, animated: true)
    }

    @IBAction func payWithSnaptokenFlow(_ sender: Any) {
        if snaptokenTextField.text?.count != 0 {
            MidtransMerchantClient.shared().requestTransacation(withCurrentToken: snaptokenTextField.text!) { (response, error) in
                if (response != nil){
                    //present payment page
                    let vc = MidtransUIPaymentViewController.init(token: response)
                    vc?.paymentDelegate = self
                    self.present(vc!, animated: true, completion: nil)
                } else {
                    print("error \(error!)");
                }
            }
        } else {
            self.showSnaptokenEmptyAlert()
        }
    }
    
    @IBAction func payWithSDKFlow(_ sender: Any) {
        
        //Create transaction details
        let itemJacket = MidtransItemDetail.init(itemID: "item001", name: "swift jacket", price: 100, quantity: 1)
        let itemPants = MidtransItemDetail.init(itemID: "item002", name: "swift pants", price: 100, quantity: 1)
        let address = MidtransAddress(firstName: "cihuy", lastName: "asik", phone: "081231231", address: "jalan buntu", city: "jakarta", postalCode: "12312", countryCode: "IDN")
        let customerDetail = MidtransCustomerDetails.init(firstName: "cihuy", lastName: "asik", email: "cihuy@cihuyhuy.com", phone: "08123123123", shippingAddress: address, billingAddress: address)
        let transactionDetail = MidtransTransactionDetails.init(orderID: self.generateRandomOrderID(), andGrossAmount: 200)

        MidtransMerchantClient.shared().requestTransactionToken(with: transactionDetail!, itemDetails: [itemJacket!, itemPants!], customerDetails: customerDetail) { (response, error) in
            if (response != nil) {
                //present payment page
                let vc = MidtransUIPaymentViewController.init(token: response)
                vc?.paymentDelegate = self
                self.present(vc!, animated: true, completion: nil)
            }
            else {
                print("error \(error!)")
            }
        }
    }

    //MARK: - MidtransUIPaymentViewControllerDelegate
    //These are the 5 mandatory transaction callback/delegates that need to be implemented to handle all the transaction status
    func paymentViewController(_ viewController: MidtransUIPaymentViewController!, paymentPending result: MidtransTransactionResult!) {
        print("payment pending")
        self.showPaymentStatusPage(paymentStatus: result.transactionStatus)
    }
    
    func paymentViewController(_ viewController: MidtransUIPaymentViewController!, paymentSuccess result: MidtransTransactionResult!) {
        print("payment success")
        self.showPaymentStatusPage(paymentStatus: result.transactionStatus)
    }
    
    func paymentViewController(_ viewController: MidtransUIPaymentViewController!, paymentFailed error: Error!) {
        print("payment failed")
        self.showPaymentStatusPage(paymentStatus: "failed")
    }
    
    func paymentViewController_paymentCanceled(_ viewController: MidtransUIPaymentViewController!) {
        print("payment cancelled")
        self.showPaymentStatusPage(paymentStatus: "cancelled")
    }
    
    func paymentViewController(_ viewController: MidtransUIPaymentViewController!, paymentDeny result: MidtransTransactionResult!) {
        print("payment denied")
        showPaymentStatusPage(paymentStatus: result.transactionStatus)
    }

}

